core/unix/loader.c

```diff
@@ -741,17 +741,55 @@ privload_search_rpath(privmod_t *mod, bool runpath, const char *name,
                     len = sep - list;
                 /* support $ORIGIN expansion to lib's current directory */
                 origin = strstr(list, RPATH_ORIGIN);
+#    ifdef CLIENT_INTERFACE
+                char path[MAXIMUM_PATH];
+#    endif
                 if (origin != NULL && origin < list + len) {
                     size_t pre_len = origin - list;
+#    ifdef CLIENT_INTERFACE
+                    snprintf(path, MAXIMUM_PATH, "%.*s%.*s%.*s", pre_len, list,
+                             moddir_len, mod->path,
+                             /* the '/' should already be here */
+                             len - strlen(RPATH_ORIGIN) - pre_len,
+                             origin + strlen(RPATH_ORIGIN));
+#    endif
                     snprintf(filename, MAXIMUM_PATH, "%.*s%.*s%.*s/%s", pre_len, list,
                              moddir_len, mod->path,
                              /* the '/' should already be here */
                              len - strlen(RPATH_ORIGIN) - pre_len,
                              origin + strlen(RPATH_ORIGIN), name);
                 } else {
+#    ifdef CLIENT_INTERFACE
+                    snprintf(path, MAXIMUM_PATH, "%.*s", len, list);
+#    endif
                     snprintf(filename, MAXIMUM_PATH, "%.*s/%s", len, list, name);
                 }
                 filename[MAXIMUM_PATH - 1] = 0;
+#    ifdef CLIENT_INTERFACE
+                NULL_TERMINATE_BUFFER(path);
+                if (mod->is_client) {
+                    /* We are adding a client's lib rpath to the general search path. This
```

Comment:
This does seem like a bandaid that is not going to solve even the same problem when it's slightly shifted: what if some other lib does the same thing but it's not the client?  libA depends on libB and libC, and libB depends on libC, and you can't find libC w/o libA's rpath, but A is not a client but pulled in by the client.  What makes the client special?  Should this *not* check for `mod->is_client`?

Reply:
Indeed. I guess the only reason I had to do this only for client libs is that I didn't want to generally do this for every system lib, as they may be most likely subject to the unhandled case of same lib name in different directories. So what makes a client lib special is the case at hand, where apparently the client library contained the system libraries rpath, but the system libraries itself don't. This is probably more common. WDYT?

Reply:
OK, if the comment makes this clear.

Reply:
Added to comment, done.